feat!: Phase checks

sirasistant opened the pull request:
The private external function macro now checks for phase changes by default. This avoids your logic to be cut in half between the setup and app logic phases when a nested call changes phase without your knowledge. If you want to opt out you need to use `#[nophasecheck]`

sirasistant commented on noir-projects/aztec-nr/aztec/src/macros/functions/mod.nr:
```diff
@@ -65,6 +65,20 @@ pub comptime fn noinitcheck(f: FunctionDefinition) {
+pub comptime fn nophasecheck(f: FunctionDefinition) {
+    // Marker attribute - see the comment above
+    if !is_fn_external(f) {
+        // Unfortunately we cannot check if we are dealing with a utility function here as Noir does not support
```
Hmmm I might need to add a check that this is not used on public functions to avoid confusion

LeilaWang commented on noir-projects/noir-protocol-circuits/crates/private-kernel-lib/src/components/reset_output_validator.nr:
```diff
@@ -254,9 +234,24 @@ impl<let NoteHashPendingReadHintsLen: u32, let NoteHashSettledReadHintsLen: u32,
+            is_revertible_requests: validation_requests.is_revertible_requests,
+            propagated_requests: output.is_revertible_requests,
+            // We use here the claimed final revertible counter, because it'll be checked in tail to match the actual one.
+            min_revertible_side_effect_counter: self.previous_kernel.claimed_revertible_counter,
```
Shouldn't this be like the split_counter that's set to 0 for private only txs?

LeilaWang replied:
Aw I guess it's difficult for the oracle resolver to know at the time a request is made if it's a private tx or not.
So an app might see that a counter is revertible, but if it ends up being in a private only tx, it's actually non-revertible. Can we think of any problem this will cause?

sirasistant replied:
Yes, I think doing the same as with squashing would make the claims that the simulator emitted fail validation, since the simulator doesn't know if the tx is going to end up private only. So for a tx that performs end setup and checks if revertible after ending setup, but ends up being private only would generate unprovable "true" claims after end setup

LeilaWang commented on noir-projects/noir-protocol-circuits/private_kernel_reset_config.json:
```diff
@@ -44,12 +44,17 @@
       "cost": 150
+    },
+    "IS_REVERTIBLE_VALIDATION": {
+      "variants": [4],
```
Maybe we can set a bigger variant here (32 or 64). Because all function calls emit at least 2 requests by default so 4 will likely not be enough. And it's not that expensive 😃

LeilaWang commented on noir-projects/aztec-nr/aztec/src/context/private_context.nr:
```diff
@@ -218,6 +219,7 @@ impl PrivateContext {
             note_hash_read_requests: BoundedVec::new(),
             nullifier_read_requests: BoundedVec::new(),
             key_validation_requests_and_generators: BoundedVec::new(),
+            is_revertible_requests: BoundedVec::new(),
```
An alternative is to have 2 numbers: a counter that's expected to be non-revertible, and a counter that's expected to be revertible.
If `in_revertible_phase` is called more than once, we override the expected value with the larger counter for non-reveretible and smaller counter for revertible.

```
pub fn in_revertible_phase(&mut self) -> bool {
        let current_counter = self.side_effect_counter;
        
        // Safety: Kernel validates that the claim is correct
        let is_revertible = unsafe {
            is_side_effect_counter_revertible_oracle_wrapper(current_counter);
        };

        if is_revertible {
            if (self.expected_revertible_counter == 0) | (current_counter < self.expected_revertible_counter) {
                self.expected_revertible_counter = current_counter;
            }
        } else if current_counter > self.expected_non_revertible_counter {
            self.expected_non_revertible_counter = current_counter;
        }

        is_revertible
    }
```

Then we check these 2 values in init and inner circuits:
```
assert((app.expected_non_revertible_counter < claimed_revertible_counter) | (claimed_revertible_counter == 0));
if app.expected_revertible_counter != 0 {
    assert((app.expected_revertible_counter >= claimed_revertible_counter) & (claimed_revertible_counter != 0));
}
```

This is slightly more expensive for the app circuits and each kernel iteration, but without the complexity of setting up variants for the reset circuits.

I'm also fine with the current implementation using reset. So if you don't see more benefits switching to this, it's all good :)

sirasistant replied:
I think I'll try to switch it, seems better!

LeilaWang reviewed (approved):
LGTM! And nice cleaning up for the hint of the claimed counter!
Just wondering what you think about the alternative?

Thunkar commented on noir-projects/noir-contracts/contracts/test/test_contract/src/main.nr:
```diff
@@ -264,6 +264,27 @@ pub contract Test {
+    #[external("private")]
+    #[nophasecheck]
+    fn call_function_that_ends_setup_without_phase_check() {
+        // This function should error since it gets inadvertently phase changed by a nested call.
```
This comment does not apply right?

sirasistant replied:
oops!

Thunkar reviewed (approved):
I'd love a TXE test for this, I believe sideEffectCounters are now correctly tracked. Otherwise LGTM

sirasistant commented:
> I'd love a TXE test for this, I believe sideEffectCounters are now correctly tracked

@Thunkar I recall TXE not allowing end setup? or am i thinking of old stuff?

Thunkar commented:
After @nventuro 's refactor I don't really see why it shouldn't. Internally it uses most of PXE's machinery and real txs are created via the same fn as kernelless simulations. Anyhow we can do it later!

sirasistant committed:
feat!: Phase checks

The private external function macro now checks for phase changes by default. This avoids your logic to be cut in half between the setup and app logic phases when a nested call changes phase without your knowledge. If you want to opt out you need to use `#[nophasecheck]`
